library/core/src/array/mod.rs

```diff
@@ -462,20 +462,47 @@ impl<T, const N: usize> [T; N] {
     ///
     /// # Note on performance and stack usage
     ///
-    /// Unfortunately, usages of this method are currently not always optimized
-    /// as well as they could be. This mainly concerns large arrays, as mapping
-    /// over small arrays seem to be optimized just fine. Also note that in
-    /// debug mode (i.e. without any optimizations), this method can use a lot
-    /// of stack space (a few times the size of the array or more).
-    ///
-    /// Therefore, in performance-critical code, try to avoid using this method
-    /// on large arrays or check the emitted code. Also try to avoid chained
-    /// maps (e.g. `arr.map(...).map(...)`).
-    ///
-    /// In many cases, you can instead use [`Iterator::map`] by calling `.iter()`
-    /// or `.into_iter()` on your array. `[T; N]::map` is only necessary if you
-    /// really need a new array of the same size as the result. Rust's lazy
-    /// iterators tend to get optimized very well.
+    /// Note that this method is *eager*.  In terms of `Iterator` methods, it's
+    /// more like `.map(…).collect()`, since it returns a new array.
+    ///
+    /// That means that `arr.map(f).map(g)` is, in general, *not* equivalent to
+    /// `array.map(|x| g(f(x)))`, as the former calls `f` 4 times then `g` 4 times,
+    /// whereas the latter interleaves the calls (`fgfgfgfg`).
```

Comment:
We shouldn't be talking about iterator evaluation order. E.g. `next_chunk` does evaluate like array.map and so any iterator method that uses chunking internally might also use that order.

Reply:
That wasn't meant to reject the update to `map`. I just meant that we shouldn't be referring to `Iterator` order since the latter depends on the implementation. No such concern applies to the array method since it's evaluated eagerly and there's no possibility of specializing that.